auth/auth.go

```diff
@@ -0,0 +1,158 @@
+// Copyright © 2018 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"encoding/base32"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/dgrijalva/jwt-go"
+	jwtRequest "github.com/dgrijalva/jwt-go/request"
+	"github.com/gin-gonic/gin"
+)
+
+type tokenStore interface {
+	Exists(userID, tokenID string) (bool, error)
+}
+
+// TokenType represents one of the possible token Types
+type TokenType string
+
+// ClaimConverter converts claims to another domain object for saving into Context
+type ClaimConverter func(*ScopedClaims) interface{}
+
+// ScopedClaims struct to store the scoped claim related things
+type ScopedClaims struct {
+	jwt.StandardClaims
+	Scope string `json:"scope,omitempty"`
+	// Drone fields
+	Type TokenType `json:"type,omitempty"`
+	Text string    `json:"text,omitempty"`
+}
+
+type options struct {
+	tokenStore tokenStore
+	extractors []jwtRequest.Extractor
+}
+
+type Option interface {
+	apply(o *options)
+}
+
+type ContextSetter func(c context.Context, value interface{}) context.Context
+
+type authOptionFunc func(o *options)
+
+func (fn authOptionFunc) apply(o *options) {
+	fn(o)
+}
+
+func TokenStoreAuthOption(tokenStore tokenStore) Option {
```

Comment:
```suggestion
func TokenStoreOption(tokenStore tokenStore) Option {
```

Since this is a dedicated auth package, maybe we can drop the Auth from the name?